src/environment/mutex.rs

```diff
@@ -0,0 +1,237 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+//! Internal mutex interface.
+
+use std::collections::HashMap;
+use std::num::NonZeroU32;
+
+use crate::libc::errno::{EBUSY, EDEADLK, EPERM};
+use crate::libc::pthread::mutex::HostMutexId;
+use crate::{Environment, ThreadID};
+
+#[derive(Default)]
+pub struct MutexState {
+    // TODO?: Maybe this should be a Vec instead? It would be bad if there were many mutexes over
+    // the lifetime of an application, but it would perform better.
+    // Maybe it could also be a fixed size allocator? (although that seems a little overkill)
+    mutexes: HashMap<HostMutexId, MutexHostObject>,
+    // Hopefully there will never be more than 2^64 mutexes in an applications lifetime :P
+    mutex_count: u64,
+}
+impl MutexState {
+    fn get_mut(env: &mut Environment) -> &mut Self {
+        &mut env.mutex_state
+    }
+    fn get(env: &Environment) -> &Self {
+        &env.mutex_state
+    }
+}
+
+struct MutexHostObject {
+    type_: MutexType,
+    waiting_count: u32,
+    /// The `NonZeroU32` is the number of locks on this thread (if it's a
+    /// recursive mutex).
+    locked: Option<(ThreadID, NonZeroU32)>,
+}
+
+#[repr(i32)]
+#[derive(Debug, PartialEq, Copy, Clone)]
+#[allow(non_camel_case_types)]
+pub enum MutexType {
+    PTHREAD_MUTEX_NORMAL = 0,
+    PTHREAD_MUTEX_ERRORCHECK = 1,
+    PTHREAD_MUTEX_RECURSIVE = 2,
+}
+
+impl TryFrom<i32> for MutexType {
+    type Error = &'static str;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(MutexType::PTHREAD_MUTEX_NORMAL),
+            1 => Ok(MutexType::PTHREAD_MUTEX_ERRORCHECK),
+            2 => Ok(MutexType::PTHREAD_MUTEX_RECURSIVE),
+            _ => Err("Value is not a valid mutex type!"),
+        }
+    }
+}
+pub const PTHREAD_MUTEX_DEFAULT: MutexType = MutexType::PTHREAD_MUTEX_NORMAL;
+/// Custom typedef for readability (the C API just uses `int`)
```

Comment:
This comment seems to have been stranded, I guess it should be removed now that you've reworked `MutexType`?

Reply:
Thanks for noticing that!